extensions/ql-vscode/src/view/results/results.tsx

```diff
@@ -164,76 +162,52 @@ export class ResultsApp extends React.Component<
 
   private updateStateWithNewResultsInfo(resultsInfo: ResultsInfo): void {
     this.setState((prevState) => {
-      const stateWithDisplayedResults = (
-        displayedResults: ResultsState,
-      ): ResultsViewState => ({
-        displayedResults,
-        isExpectingResultsUpdate: prevState.isExpectingResultsUpdate,
-        nextResultsInfo: resultsInfo,
-      });
-
-      if (!prevState.isExpectingResultsUpdate && resultsInfo === null) {
-        // No results to display
-        return stateWithDisplayedResults({
-          resultsInfo: null,
-          results: null,
-          errorMessage: "No results to display",
-        });
-      }
-      if (!resultsInfo || !resultsInfo.shouldKeepOldResultsWhileRendering) {
-        // Display loading message
-        return stateWithDisplayedResults({
-          resultsInfo: null,
-          results: null,
-          errorMessage: "Loading results…",
-        });
+      if (resultsInfo === null) {
+        if (prevState.isExpectingResultsUpdate) {
+          // Display loading message
+          return {
+            displayedResults: {
+              resultsInfo: null,
+              results: null,
+              errorMessage: "Loading results…",
+            },
+            isExpectingResultsUpdate: prevState.isExpectingResultsUpdate,
+            nextResultsInfo: resultsInfo,
+          };
+        } else {
+          // No results to display
+          return {
+            displayedResults: {
+              resultsInfo: null,
+              results: null,
+              errorMessage: "No results to display",
+            },
+            isExpectingResultsUpdate: prevState.isExpectingResultsUpdate,
+            nextResultsInfo: resultsInfo,
+          };
+        }
```

Comment:
Are you missing a case here? I don't see line 191 from the original handled.

```
return stateWithDisplayedResults(prevState.displayedResults);
```

Reply:
In this case, the missing case is correct. In the previous code, after the `setState` returned with `stateWithDisplayedResults(prevState.displayedResults)`, `this.loadResults()` would be called immediately. The only reason for storing the `nextResultsInfo` in the state is so that `loadResults` can use it. However, with concurrent updates the state seen in `loadResults` would not yet be the state set in this `setState` call. Therefore, I've combined this `setState` and the `loadResults` method. Therefore, this case won't happen anymore.

Thanks for the suggestion for the simplification. Since we don't need the third case, I couldn't use it as-is. However, I have split the flows differently so there's only 1 level of nesting rather than 2.